add.go

```diff
@@ -0,0 +1,231 @@
+package ipfscluster
+
+import (
+	"context"
+	"errors"
+	"mime/multipart"
+	"net/url"
+	"strconv"
+
+	"github.com/ipfs/ipfs-cluster/api"
+	"github.com/ipfs/ipfs-cluster/importer"
+
+	"github.com/ipfs/go-ipfs-cmdkit/files"
+)
+
+func (c *Cluster) consumeLocalAdd(
+	args map[string]string,
+	outObj *api.NodeWithMeta,
+	replMin, replMax int,
+) error {
+	//TODO: when ipfs add starts supporting formats other than
+	// v0 (v1.cbor, v1.protobuf) we'll need to update this
+	outObj.Format = ""
+	args["cid"] = outObj.Cid // root node stored on last call
+	var hash string
+	err := c.rpcClient.Call(
+		"",
+		"Cluster",
+		"IPFSBlockPut",
+		*outObj,
+		&hash)
+	if outObj.Cid != hash {
+		logger.Warningf("mismatch. node cid: %s\nrpc cid: %s", outObj.Cid, hash)
+	}
+	return err
+}
+
+func (c *Cluster) finishLocalAdd(
+	args map[string]string,
+	replMin, replMax int,
+) error {
+	rootCid, ok := args["cid"]
+	if !ok {
+		return errors.New("no record of root to pin")
+	}
+
+	pinS := api.PinSerial{
+		Cid:                  rootCid,
+		Type:                 api.DataType,
+		Recursive:            true,
+		ReplicationFactorMin: replMin,
+		ReplicationFactorMax: replMax,
+	}
+	return c.rpcClient.Call(
+		"",
+		"Cluster",
+		"Pin",
+		pinS,
+		&struct{}{},
+	)
+}
+
+func (c *Cluster) consumeShardAdd(
+	args map[string]string,
+	outObj *api.NodeWithMeta,
+	replMin, replMax int,
+) error {
+
+	var shardID string
+	shardID, ok := args["id"]
+	outObj.ID = shardID
+	outObj.ReplMax = replMax
+	outObj.ReplMin = replMin
+	var retStr string
+	err := c.rpcClient.Call(
+		"",
+		"Cluster",
+		"SharderAddNode",
+		*outObj,
+		&retStr)
+	if !ok {
+		args["id"] = retStr
+	}
+	return err
+}
+
+func (c *Cluster) finishShardAdd(
+	args map[string]string,
+	replMin, replMax int,
+) error {
+	shardID, ok := args["id"]
+	if !ok {
+		return errors.New("bad state: shardID passed incorrectly")
+	}
+	err := c.rpcClient.Call(
+		"",
+		"Cluster",
+		"SharderFinalize",
+		shardID,
+		&struct{}{},
+	)
+	return err
+}
+
+func (c *Cluster) consumeImport(ctx context.Context,
+	outChan <-chan *api.NodeWithMeta,
+	printChan <-chan *api.AddedOutput,
+	errChan <-chan error,
+	consume func(map[string]string, *api.NodeWithMeta, int, int) error,
+	finish func(map[string]string, int, int) error,
+	replMin int, replMax int,
+) ([]api.AddedOutput, error) {
+	var err error
+	openChs := 3
+	toPrint := make([]api.AddedOutput, 0)
+	args := make(map[string]string)
+
+	for {
+		if openChs == 0 {
+			break
+		}
+
+		// Consume signals from importer.  Errors resulting from
+		select {
+		// Ensure we terminate reading from importer after cancellation
+		// but do not block
+		case <-ctx.Done():
+			err = errors.New("context timeout terminated add")
+			return nil, err
+		// Terminate session when importer throws an error
+		case err, ok := <-errChan:
+			if !ok {
+				openChs--
+				errChan = nil
```

Comment:
Why do you set the `channels` to `nil`, if they have been closed by another goroutine? I am curious more than anything else.

Reply:
Good question.  My understanding is that closed channels are always available, and nil channels never are.  The `openChs` count requires listening on a closed channel exactly once.  See https://stackoverflow.com/questions/13666253/breaking-out-of-a-select-statement-when-all-channels-are-closed?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa for more.